modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +544,172 @@ export class OlmAdapter extends Listenable {
+                    logger.warn('SAS already has their key!');
+
+                    return;
+                }
+
+                const { key: theirKey, isInitializer, uuid } = msg.data;
+
+                if (olmData.sasCommitment) {
+                    const olmUtil = new Olm.Utility();
+                    const commitment = olmUtil.sha256(theirKey + olmData.startContent);
+                    olmUtil.free();
+
+                    if (olmData.sasCommitment !== commitment) {
+                        throw new Error('OlmAdapter commitments mismatched');
+                    }
+                }
+
+                olmData.sas.set_their_key(theirKey);
+
+                const pub_key = olmData.sas.get_pubkey();
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pub_key,
+                            uuid
+                        }
+                    }
+                };
+                this._sendMessage(ack, pId);
+
+                const myInfo = `${this.myId}|${pub_key}`;
+                const theirInfo = `${pId}|${theirKey}`
+
+                const info = isInitializer ? `${myInfo}|${theirInfo}` : `${theirInfo}|${myInfo}`;
+
+                const sasBytes = olmData.sas.generate_bytes(info, OLM_SAS_NUM_BYTES);
+                const sas = generateSas(sasBytes);
+
+                this.eventEmitter.emit(OlmAdapterEvents.PARTICIPANT_SAS_READY, pId, sas);
+            } else {
+                logger.debug(`Received sas key message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-key');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_MAC: {
+            if (olmData.session) {
+                const { keys, mac, uuid } = msg.data;
+
+                if (!mac || !keys) {
+                    logger.warn('Invalid SAS MAC message');
+                    return;
+                }
+
+                // Verify the received MACs.
+                const baseInfo = `${OLM_KEY_VERIFICATION_MAC_INFO}${pId}${this.myId}${uuid}`;
+                const keysMac = olmData.sas.calculate_mac(
+                    Object.keys(mac).sort().join(','), // eslint-disable-line newline-per-chained-call
+                    baseInfo + OLM_KEY_VERIFICATION_MAC_KEY_IDS,
+                );
+
+                if (keysMac !== keys) {
+                    logger.error('SAS verification error: keys MAC mismatch');
+                    this.eventEmitter.emit(OlmAdapterEvents.PARTICIPANT_SAS_COMPLETED, pId, false);
+                    return;
+                }
+
+                for (const [ keyInfo, computedMac ] of Object.entries(mac)) {
+                    const keyType = keyInfo.split(':')[0];
+                    const pubKey = olmData.lastKey;
```

Comment:
This should be the target participant's ed25519 key which you need to learn it from presence.